Not compatible with MySQL 8.0

MySQL 8.0 has introduced cache to improve the efficiency of information_schema queries, so it may lead to a consequence that users could not get the latest and accurate statistics when querying information_schema.

If using MySQL 8.0 as storage layer, trillian codes, without setting the session variable `information_schema_stats_expiry` (the default is 86400 seconds), could not pass the `TestMySQLRateLimiting` and `TestQuotaManager_GetTokens_InformationSchema`. 
```
--- FAIL: TestMySQLRateLimiting (1.13s)
    quota_test.go:100: Test MySQL available at "root@tcp(127.0.0.1)/"
    quota_test.go:124: timed out before rate limiting kicked in
FAIL
FAIL    github.com/google/trillian/integration/quota    8.509s
```

```
--- FAIL: TestQuotaManager_GetTokens_InformationSchema (1.44s)
    mysql_quota_test.go:118: Test MySQL available at "root@tcp(127.0.0.1)/"
    --- FAIL: TestQuotaManager_GetTokens_InformationSchema/useSelectCount_=_false (1.34s)
        mysql_quota_test.go:170: timed out
FAIL
FAIL    github.com/google/trillian/quota/mysqlqm        4.407s
```


The countFromInformationSchema function in the codes depends on information_schema to return the latest statistics,
```golang
	countFromInformationSchemaQuery = `
		SELECT table_rows
		FROM information_schema.tables
		WHERE table_schema = schema()
			AND table_name = ?
			AND table_type = ?`
```

```golang
func countFromInformationSchema(ctx context.Context, db *sql.DB) (int, error) {
	// information_schema.tables doesn't have an explicit PK, so let's play it safe and ensure
	// the cursor returns a single row.
	rows, err := db.QueryContext(ctx, countFromInformationSchemaQuery, "Unsequenced", "BASE TABLE")
	if err != nil {
		return 0, err
	}
	defer rows.Close()
	if !rows.Next() {
		return 0, errors.New("cursor has no rows after information_schema query")
	}
	var count int
	if err := rows.Scan(&count); err != nil {
		return 0, err
	}
	if rows.Next() {
		return 0, errors.New("too many rows returned from information_schema query")
	}
	return count, nil
}
```

so anything that relies on this function may fail to work as normal.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
   `merkle` libraries in this repository. This is not a breaking change, but
   we recommend clients also migrate over to this library at the earliest
   convenient time; the long term plan is to remove `merkle` from this repo.
+* `countFromInformationSchema` function to add support for MySQL 8.
 
 ## v1.4.0
 

--- a/quota/mysqlqm/mysql_quota.go
+++ b/quota/mysqlqm/mysql_quota.go
@@ -19,6 +19,8 @@ import (
 	"context"
 	"database/sql"
 	"errors"
+	"fmt"
+	"strconv"
 
 	"github.com/google/trillian/quota"
 )
@@ -97,6 +99,10 @@ func (m *QuotaManager) countUnsequenced(ctx context.Context) (int, error) {
 }
 
 func countFromInformationSchema(ctx context.Context, db *sql.DB) (int, error) {
+	// turn off statistics caching for MySQL 8
+	if err := turnOffInformationSchemaCache(ctx, db); err != nil {
+		return 0, err
+	}
 	// information_schema.tables doesn't have an explicit PK, so let's play it safe and ensure
 	// the cursor returns a single row.
 	rows, err := db.QueryContext(ctx, countFromInformationSchemaQuery, "Unsequenced", "BASE TABLE")
@@ -123,4 +129,30 @@ func countFromTable(ctx context.Context, db *sql.DB) (int, error) {
 		return 0, err
 	}
 	return count, nil
+}
+
+// turnOffInformationSchemaCache turn off statistics caching for MySQL 8
+// To always retrieve the latest statistics directly from the storage engine and bypass cached values, set information_schema_stats_expiry to 0.
+// See https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_information_schema_stats_expiry
+func turnOffInformationSchemaCache(ctx context.Context, db *sql.DB) error {
+	opt := "information_schema_stats_expiry"
+	res := db.QueryRowContext(ctx, "SHOW VARIABLES LIKE '%"+opt+"%'")
+	var none, expiry string
+	err := res.Scan(&none, &expiry)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil
+		}
+
+		return fmt.Errorf("Failed to get variable %q: %v", opt, err)
+	}
+
+	exp, err := strconv.Atoi(expiry)
+	if err != nil || exp != 0 {
+		if _, err := db.ExecContext(ctx, "SET SESSION "+opt+"=0"); err != nil {
+			return fmt.Errorf("Failed to set variable %q: %v", opt, err)
+		}
+	}
+
+	return nil
 }

--- a/quota/mysqlqm/mysql_quota.go
+++ b/quota/mysqlqm/mysql_quota.go
@@ -20,7 +20,6 @@ import (
 	"database/sql"
 	"errors"
 	"fmt"
-	"strconv"
 
 	"github.com/google/trillian/quota"
 )
@@ -134,23 +133,25 @@ func countFromTable(ctx context.Context, db *sql.DB) (int, error) {
 // turnOffInformationSchemaCache turn off statistics caching for MySQL 8
 // To always retrieve the latest statistics directly from the storage engine and bypass cached values, set information_schema_stats_expiry to 0.
 // See https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_information_schema_stats_expiry
+// MySQL versions prior to 8 will fail safely.
 func turnOffInformationSchemaCache(ctx context.Context, db *sql.DB) error {
 	opt := "information_schema_stats_expiry"
-	res := db.QueryRowContext(ctx, "SHOW VARIABLES LIKE '%"+opt+"%'")
-	var none, expiry string
-	err := res.Scan(&none, &expiry)
-	if err != nil {
+	res := db.QueryRowContext(ctx, "SHOW VARIABLES LIKE '"+opt+"'")
+	var none string
+	var expiry int
+
+	if err := res.Scan(&none, &expiry); err != nil {
+		// fail safely for all versions of MySQL prior to 8
 		if errors.Is(err, sql.ErrNoRows) {
 			return nil
 		}
 
-		return fmt.Errorf("Failed to get variable %q: %v", opt, err)
+		return fmt.Errorf("failed to get variable %q: %v", opt, err)
 	}
 
-	exp, err := strconv.Atoi(expiry)
-	if err != nil || exp != 0 {
+	if expiry != 0 {
 		if _, err := db.ExecContext(ctx, "SET SESSION "+opt+"=0"); err != nil {
-			return fmt.Errorf("Failed to set variable %q: %v", opt, err)
+			return fmt.Errorf("failed to set variable %q: %v", opt, err)
 		}
 	}
 

